traittypes/traittypes.py

```diff
@@ -70,13 +85,20 @@ class Array(SciType):
     def validate(self, obj, value):
         if value is None and not self.allow_none:
             self.error(obj, value)
+        if value is None or value is Undefined:
+            return super(Array, self).validate(obj, value)
         try:
-            value = np.asarray(value, dtype=self.dtype)
-            for validator in self.validators:
-                value = validator(self, value)
-            return value
+            r = np.asarray(value, dtype=self.dtype)
+            if isinstance(value, np.ndarray) and r is not value:
+                warnings.warn(
+                    'Given trait value dtype "%s" does not match required type "%s". '
+                    'A coerced copy has been created.' % (
+                        np.dtype(value.dtype).name,
+                        np.dtype(self.dtype).name))
```

Comment:
Do you want this warning to occur if you do `foo.bar = [1, 2]`?

Reply:
I didn't think so (hence the test if original value is an ndarray). Do you think it should?

Reply:
Ah gotcha.

Reply:
No, I agree with you.